### ISSUE TITLE
Non ruby/passenger entries populating passenger error log

### ISSUE THREAD

**Issue by ylluminate** [phusion/passenger#1373]
I am curious, I am noticing a lot of non-ruby traffic populating the "ruby_error_log" that I have instructed `mod_passenger.c` to keep via:
```
PassengerDebugLogFile /usr/local/apache/logs/ruby_error_log
```

The entries are many and varied such as:
```
[tag "OWASP_CRS/PROTOCOL_VIOLATION/MISSING_HEADER_ACCEPT"] [tag "WASCTC/WASC-21"] [tag "OWASP_TOP_10/A7"] [tag "PCI/6.5.10"] [hostname "redacteddomain.tld"] [uri "/"] [unique_id "VNt7-MZkddoABuQClmUAAAA7"]
[Wed Feb 11 10:57:49.074166 2015] [:error] [pid 451586] [client 144.76.32.148] ModSecurity: Access denied with redirection to http://redacteddomain.tld/ using status 302 (phase 2). Match of "pm AppleWebKit Android" against "REQUEST_HEADERS:User-Agent" required. [file "/usr/l
[Wed Feb 11 10:58:07.274498 2015] [fcgid:warn] [pid 451419] [client 157.55.39.13:56954] mod_fcgid: read data timeout in 40 seconds
[Wed Feb 11 10:58:07.274567 2015] [core:error] [pid 451419] [client 157.55.39.13:56954] End of script output before headers: index.php
[Wed Feb 11 10:58:08.341603 2015] [:error] [pid 451572] [client 123.125.71.86] ModSecurity: Rule processing failed. [hostname "www.redacteddomain.tld"] [uri "/some-ver-long-paths-appear-here-sometimes-frequently/"] [unique_id "VNt8EMZkddoABuP0@YYAAAAt"]
[Wed Feb 11 11:00:26.571884 2015] [:error] [pid 455729] [client 72.50.229.162] ModSecurity: Rule processing failed. [hostname "redacteddomain.tld"] [uri "/feed/"] [unique_id "VNt8msZkddoABvQxavkAAAAA"]
```

Any idea as to why these are appearing here instead of solely within the apache `error_log`?

**Comment by FooBarWidget** [comment]
It should not be possible for Passenger to redirect other Apache modules' output to its log file. Are you sure you didn't configure Apache to log errors/requests to that file too?

**Comment by ylluminate** [comment]
I am as sure as I can be at present.  After adding this config within initially to get Passenger operational some time ago `/usr/local/apache/conf/includes/pre_main_global.conf`:
```
LoadModule passenger_module /usr/local/rvm/gems/ruby-2.1.5/gems/passenger-5.0.0.beta3/buildout/apache2/mod_passenger.so
<IfModule mod_passenger.c>
  PassengerRoot /usr/local/rvm/gems/ruby-2.1.5/gems/passenger-5.0.0.beta3
  PassengerDefaultRuby /usr/local/rvm/gems/ruby-2.1.5/wrappers/ruby
  PassengerLogLevel 2
  PassengerDebugLogFile /usr/local/apache/logs/ruby_error_log
</IfModule>

<IfModule mod_reqtimeout.c>
   RequestReadTimeout header=20-40,MinRate=500 body=20,MinRate=500
</IfModule>

<IfModule mod_proxy.c>
  ProxyRequests Off
  <Proxy *>
    AddDefaultCharset off
    Order allow,deny
    Allow from all
  </Proxy>
  ProxyVia On
  ProxyPreserveHost On
</IfModule>
```

I began getting a lot of these entries within the `ruby_error_log` instead of the apache `error_log`.  Am I passing a wrong configuration argument here to accomplish this?

**Comment by FooBarWidget** [comment]
Your configuration is fine. I haven't been able to reproduce this though.

**Comment by ylluminate** [comment]
Hmm. It is quite a headache since the separation is not happening as it normally should.  I've asked cPanel support to chime in, but I suspect they'll push it away since it is "out of scope" given the inclusion of Passenger in the mix.

The ruby log has become so polluted now with a recent update from them to the web app firewall system that I simply cannot use it at all anymore to troubleshoot Passenger app issues.

`httpd -V` yields:
```
Server version: Apache/2.4.10 (Unix)
Server built:   Jan 22 2015 16:43:52
Cpanel::Easy::Apache v3.28.1 rev9999 +cloudlinux
Server's Module Magic Number: 20120211:36
Server loaded:  APR 1.5.1, APR-UTIL 1.5.4
Compiled using: APR 1.5.1, APR-UTIL 1.5.4
Architecture:   64-bit
Server MPM:     prefork
  threaded:     no
    forked:     yes (variable process count)
Server compiled with....
 -D APR_HAS_SENDFILE
 -D APR_HAS_MMAP
 -D APR_HAVE_IPV6 (IPv4-mapped addresses disabled)
 -D APR_USE_SYSVSEM_SERIALIZE
 -D APR_USE_PTHREAD_SERIALIZE
 -D SINGLE_LISTEN_UNSERIALIZED_ACCEPT
 -D APR_HAS_OTHER_CHILD
 -D AP_HAVE_RELIABLE_PIPED_LOGS
 -D DYNAMIC_MODULE_LIMIT=256
 -D HTTPD_ROOT="/usr/local/apache"
 -D SUEXEC_BIN="/usr/local/apache/bin/suexec"
 -D DEFAULT_PIDLOG="logs/httpd.pid"
 -D DEFAULT_SCOREBOARD="logs/apache_runtime_status"
 -D DEFAULT_ERRORLOG="logs/error_log"
 -D AP_TYPES_CONFIG_FILE="conf/mime.types"
 -D SERVER_CONFIG_FILE="conf/httpd.conf"
```

So as you can note the logging for apache proper should be going to the noted error_log file at HTTPD_ROOT/logs/error_log vs the ruby_error_log as I defined.  Something must be causing Passenger, once loaded, to redirect log events that are higher than `:warn` and `:notice` entries over to the ruby_error_log because when I look in it I primarily see apache `:error` level entries as well as, of course, ruby / Passenger entries...

Not sure if helpful, but here are the compiled in modules for apache as well:
https://gist.github.com/0f3b4b13897581226bfa

**Comment by ylluminate** [comment]
Well @FooBarWidget, I received the following back from cPanel support:
```
I have been watching the logs ruby_error_log and modsec_audit.log populate and both contain identical errors, though both are formatted differently. This suggests that the behavior you are seeing is actually normal.

It looks at a glance as though Passenger is picking up on the warning messages provided from ModSecurity, etc., and then logging those messages in it's own manner. Unfortunately I am unfamiliar with Passenger and it's logging, and it appears that the processes are spun up and down dynamically making it nearly impossible to strace them accurately.
```

**Comment by ylluminate** [comment]
He added a little bit of additional interesting information:
> I wasn't able to see any apache specific log information in the ruby_error_log, though I did see information that is typically logged by apache, however all of the log information I saw seemed to have been processed by Passenger's logging mechanisms which suggests that it is hooking directly into the apache logs to get it's information, but is not filtering what information it takes from those logs; It simply takes all of the information, based on the debug level configured, parses and logs it.

**Comment by ylluminate** [comment]
Any more thoughts here @FooBarWidget?

**Comment by FooBarWidget** [comment]
What he said in the last comment is not true. We do not hook into Apache's logging API, nor do we capture any Apache logs to re-log it later.

Our method of logging is as follows and happens from multiple places:

1. When Apache is started (in the init_module phase), we fork a process. This subprocess opens the passenger_debug_log_file and attaches it to stdout and stderr. All logs are then written to stderr.
2. Inside the Apache module, we log by writing directly to the passenger_debug_log_file.

Nothing fancy going on here.

**Comment by ylluminate** [comment]
An FYI, I have related all of this to cPanel for further research.  They are working on setting up a duplicate environment to see where the problem is coming into play here because they are concerned since this should not be happening and seems to perhaps coincide with an update.

**Comment by ylluminate** [comment]
FYI @FooBarWidget, cPanel is currently attempting to reproduce the issue so that their engineering department and isolate the issue here.  They are having some challenges it appears even after studying my actual setup.

**Comment by FooBarWidget** [comment]
I found the problem. Inside the Apache module, Passenger calls dup2() to redirect stdout and stderr to the log file. As a result, anything that Apache and its other modules write to stdout and stderr are redirected to the Passenger log file. I'll schedule the fix for 5.0.6.

**Comment by ylluminate** [comment]
:heart:

**Comment by FooBarWidget** [comment]
Fixed in 5.0.11.
